waddrmgr/manager.go

```diff
@@ -963,6 +973,147 @@ func (m *Manager) existsAddress(addressID []byte) (bool, error) {
 	return exists, nil
 }
 
+// ImportPublicKey imports a raw pubkey as a watch-only address into the address manager.
+//
+// All imported addresses will be part of the account defined by the
+// ImportedAddrAccount constant.
+func (m *Manager) ImportPublicKey(pubKey *btcec.PublicKey, bs *BlockStamp) (ManagedPubKeyAddress, error) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	serializedPubKey := pubKey.SerializeCompressed()
+	pubKeyHash := btcutil.Hash160(serializedPubKey)
+	alreadyExists, err := m.existsAddress(pubKeyHash)
+	if err != nil {
+		return nil, err
+	}
+	if alreadyExists {
+		str := fmt.Sprintf("address for public key %x already exists",
+			serializedPubKey)
+		return nil, managerError(ErrDuplicateAddress, str, nil)
+	}
+
+	// Encrypt public key.
+	encryptedPubKey, err := m.cryptoKeyPub.Encrypt(serializedPubKey)
+	if err != nil {
+		str := fmt.Sprintf("failed to encrypt public key for %x",
+			serializedPubKey)
+		return nil, managerError(ErrCrypto, str, err)
+	}
+
+	// The start block needs to be updated when the newly imported address
+	// is before the current one.
+	updateStartBlock := bs.Height < m.syncState.startBlock.Height
+
+	// Save the new imported address to the db and update start block (if
+	// needed) in a single transaction.
+	err = m.namespace.Update(func(tx walletdb.Tx) error {
+		err := putImportedAddress(tx, pubKeyHash, ImportedAddrAccount,
+			ssNone, true, nil, encryptedPubKey, nil)
+		if err != nil {
+			return err
+		}
+
+		if updateStartBlock {
+			return putStartBlock(tx, bs)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Now that the database has been updated, update the start block in
+	// memory too if needed.
+	if updateStartBlock {
+		m.syncState.startBlock = *bs
+	}
+
+	// Create a new managed address based on the imported address.
+	var managedAddr *managedAddress
+	managedAddr, err = newManagedAddressWithoutPrivKey(m,
+		ImportedAddrAccount, pubKey, true)
+	if err != nil {
+		return nil, err
+	}
+	managedAddr.imported = true
+
+	// Add the new managed address to the cache of recent addresses and
+	// return it.
+	m.addrs[addrKey(managedAddr.Address().ScriptAddress())] = managedAddr
+	return managedAddr, nil
+}
+
+// ImportAddress imports a raw pkhash as a watch-only address into the address manager.
+//
+// All imported addresses will be part of the account defined by the
+// ImportedAddrAccount constant.
+func (m *Manager) ImportAddress(addr btcutil.Address, bs *BlockStamp) (ManagedAddress, error) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	pubKeyHash := addr.ScriptAddress()
```

Comment:
As mentioned below, this function should be looking at the type of the address with a type assertion and handling it appropriately for all supported types.  See my comment in the Wallet's `ImportAddress` function for more details.